clippy_lints/src/question_mark.rs

```diff
@@ -109,12 +109,31 @@ fn find_let_else_ret_expression<'hir>(block: &'hir Block<'hir>) -> Option<&'hir
 }
 
 fn check_let_some_else_return_none(cx: &LateContext<'_>, stmt: &Stmt<'_>) {
+    /// Make sure the init expr implements try trait so a valid suggestion could be given.
+    ///
+    /// Because the init expr could have the type of `&Option<T>` which does not implements `Try`.
+    ///
+    /// NB: This conveniently prevents the cause of
+    /// issue [#12412](https://github.com/rust-lang/rust-clippy/issues/12412),
+    /// since accessing an `Option` field from a borrowed struct requires borrow, such as
+    /// `&some_struct.opt`, which is type of `&Option`. And we can't suggest `&some_struct.opt?`
+    /// or `(&some_struct.opt)?` since the first one has different semantics and the later does
+    /// not implements `Try`.
+    fn init_expr_can_use_question_mark(cx: &LateContext<'_>, init_expr: &Expr<'_>) -> bool {
+        let init_ty = cx.typeck_results().expr_ty_adjusted(init_expr);
+        cx.tcx
+            .lang_items()
+            .try_trait()
+            .map_or(false, |did| implements_trait(cx, init_ty, did, &[]))
+    }
```

Comment:
Does this pattern come up often? I think I have seen something like this before. Perhaps we could have this in the utils mod.

Reply:
`cx.tcx.lang_items.something().map_or(false, |id| some_util(...))` comes up fairly often, some utils have different variants that take lang items or diagnostic names, maybe we could do something like

```rust
trait Definition {
    fn def_id(&self) -> Option<DefId>;
}
```

And impl it for `DefId`, `LangItem`, `Symbol` (for diagnostic names)

Then you could do `implements_trait(..., LangItem::Try)` or `implements_trait(.., sym::Default)`. Could use it to get unify a good few utils, but things taking a trait is going to be slightly more difficult to understand